File: backend/app/schemas/device.py
from pydantic import BaseModel
from typing import Optional, List, Any
from datetime import datetime
from app.schemas.content import ContentResponse
from app.schemas.zone import ZoneResponse
from app.schemas.fleet import FleetResponse

class DeviceBase(BaseModel):
    device_id: str
    name: str
    device_type: Optional[str] = "BUS"  # BUS, TRAIN, STATION, DIGITAL_SIGNAGE, KIOSK, LED_SCREEN
    fleet_id: Optional[int] = None
    location_name: Optional[str] = None
    refresh_interval: Optional[int] = 10
    default_content_id: Optional[int] = None
    override_content_id: Optional[int] = None

class DeviceCreate(DeviceBase):
    token: Optional[str] = None

class DeviceUpdate(BaseModel):
    name: Optional[str] = None
    device_type: Optional[str] = None
    fleet_id: Optional[int] = None
    location_name: Optional[str] = None
    refresh_interval: Optional[int] = None
    default_content_id: Optional[int] = None
    override_content_id: Optional[int] = None

class DeviceLocationUpdate(BaseModel):
    latitude: float
    longitude: float

class DeviceHeartbeat(BaseModel):
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    timestamp: Optional[datetime] = None

class DeviceResponse(DeviceBase):
    id: int
    token: str
    current_lat: Optional[float] = None
    current_lon: Optional[float] = None
    location_name: Optional[str] = None
    current_zone_id: Optional[int] = None
    status: str  # "ONLINE", "OFFLINE", "WARNING"
    last_seen: Optional[datetime] = None
    created_at: datetime
    active_content: Optional[ContentResponse] = None
    current_zone: Optional[ZoneResponse] = None
    fleet: Optional[FleetResponse] = None
    content_reason: Optional[str] = None

    class Config:
        from_attributes = True

class DeviceStats(BaseModel):
    total_devices: int
    online_devices: int
    offline_devices: int
    warning_devices: int
    active_zones: int

class DevicePaginatedResponse(BaseModel):
    items: List[DeviceResponse]
    total: int
    page: int
    page_size: int
    total_pages: int
    stats: DeviceStats

class DeviceLogItem(BaseModel):
    id: int
    timestamp: datetime
    event_type: str
    message: Optional[str] = None
    details_json: Optional[str] = None

    class Config:
        from_attributes = True

class DeviceDetailResponse(DeviceResponse):
    assignment_reason: Optional[str] = None
    assigned_by: Optional[str] = None
    recent_logs: List[DeviceLogItem] = []

class ContentPayload(BaseModel):
    id: int
    title: str
    file_url: str
    media_type: str = "image"
    duration: Optional[float] = 10.0
    tags: Optional[str] = None

class DeviceCurrentContentResponse(BaseModel):
    device_id: str
    content_id: Optional[int] = None
    content: Optional[ContentPayload] = None
    reason: str  # e.g., "ZONE_SCHEDULE", "ZONE_ASSIGNMENT", "DEVICE_OVERRIDE", "DEFAULT_FALLBACK", "NONE"
    zone_id: Optional[int] = None
    zone_name: Optional[str] = None
    timestamp: datetime = datetime.utcnow()

class DeviceLocationMarker(BaseModel):
    device_id: str
    name: str
    device_type: Optional[str] = "BUS"
    fleet_name: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    location_name: Optional[str] = None
    status: str
    current_zone: Optional[str] = None
    current_content_title: Optional[str] = None
    current_content_url: Optional[str] = None
    last_seen: Optional[datetime] = None

class DeviceActionRequest(BaseModel):
    action: str  # "restart_player", "sync_device", "remote_config"
    params: Optional[dict] = None

File: frontend/src/types/device.ts
export type DeviceType =
  | 'BUS'
  | 'TRAIN'
  | 'STATION'
  | 'DIGITAL_SIGNAGE'
  | 'KIOSK'
  | 'LED_SCREEN';

export type DeviceStatus = 'ONLINE' | 'OFFLINE' | 'WARNING';

export interface Fleet {
  id: number;
  name: string;
  organization?: string;
  device_type: string;
  created_at: string;
  device_count?: number;
}

export interface DeviceContent {
  id: number;
  title: string;
  description?: string;
  file_url: string;
  media_type?: string;
  duration?: number;
}


export interface DeviceListItem {
  id: number;
  device_id: string;
  name: string;
  device_type: DeviceType;
  fleet_id?: number;
  fleet?: Fleet;
  location_name?: string;
  current_lat?: number;
  current_lon?: number;
  current_zone_id?: number;
  current_zone?: {
    id: number;
    name: string;
    zone_type: string;
    color?: string;
  };
  active_content?: DeviceContent;
  status: DeviceStatus;
  last_seen?: string;
  created_at: string;
  refresh_interval: number;
  token: string;
}

export interface DeviceStats {
  total_devices: number;
  online_devices: number;
  offline_devices: number;
  warning_devices: number;
  active_zones: number;
}

export interface DevicePaginatedResponse {
  items: DeviceListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
  stats: DeviceStats;
}

export interface DeviceLogItem {
  id: number;
  timestamp: string;
  event_type: string;
  message?: string;
  details_json?: string;
}

export interface PlaylistItem {
  campaign_id?: number | null;
  content_id?: number | null;
  title: string;
  file_url?: string | null;
  media_type?: string;
  duration?: number;
  priority?: number;
  description?: string | null;
}

export interface DeviceDetail extends DeviceListItem {
  assignment_reason?: string;
  assigned_by?: string;
  recent_logs: DeviceLogItem[];
  override_content_id?: number;
  default_content_id?: number;
  playlist?: PlaylistItem[];
  slot_duration?: number;
}

export interface DeviceQueryParams {
  page?: number;
  page_size?: number;
  search?: string;
  device_type?: string;
  fleet_id?: number;
  zone_id?: number;
  status?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
}

export interface DeviceCreateInput {
  device_id: string;
  name: string;
  device_type: DeviceType;
  fleet_id?: number;
  initial_zone_id?: number;
  token?: string;
  refresh_interval?: number;
}

// ─── Location Marker (for live map) ─────────────────────────────────────────
export interface DeviceLocationMarker {
  device_id: string;
  name: string;
  device_type?: string;
  fleet_name?: string;
  latitude?: number | null;
  longitude?: number | null;
  location_name?: string;
  status: DeviceStatus;
  current_zone?: string | null;
  current_content_title?: string | null;
  current_content_url?: string | null;
  playlist?: PlaylistItem[];
  slot_duration?: number;
  last_seen?: string | null;
}

// ─── Legacy aliases (for backward compatibility) ──────────────────────────────
export type Device = DeviceListItem;
export type DeviceCreate = DeviceCreateInput;
export interface DeviceUpdate {
  name?: string;
  description?: string;
  override_content_id?: number;
  fallback_content_id?: number;
  refresh_interval?: number;
}
